Add ipsec-client 3.4.0 release notes (#411)
128technology/docs: docs/plugin_ipsec_client.md

## Configuration

> The IPsec plugin setup has the following key parts to the configuration

Reviewer: ```suggestion
The IPsec plugin setup has the following key parts to the configuration.
```

> * `ipsec-profile` describing the mechanism with which to connect to the server

Reviewer: ```suggestion
* `ipsec-profile` describing the mechanism with which to connect to the server.
```

> * `ipsec-client` represent the remote endpoints or server that the ipsec client communicates with

Reviewer: ```suggestion
* `ipsec-client` represent the remote endpoints or server with which the ipsec client communicates.
```

* `service-route`'s to route the traffic through the tunnels

### Profiles

> The `router > ipsec-profile`'s are reusable IPsec settings that can be used across multiple nodes in a router and multiple IPsec endpoint `remote`s. The table below represents the most common configuration elements for a valid ipsec profile

Reviewer: ```suggestion
The `router > ipsec-profile`'s are reusable IPsec settings that can be used across multiple nodes in a router and multiple IPsec endpoint `remote`s. The table below represents the most common configuration elements for a valid ipsec profile.
```

:::

### Clients
Clients are a collection of remote endpoints which can be used for connection and failover purposes.

The main config properties of a remote endpoint are as follows.

| Config            | Type      | Description                            |
| --------          | -----     | -------------------                    |
| name              | string    | The name of the remote client to be used for sending traffic to the tunnel. |

> | host              | ip-or-fqdn | The address or FQDN of the remote endpoint |

Reviewer: ```suggestion
| host              | ip-or-fqdn | The address or FQDN of the remote endpoint. |
```

> | profile           | reference | The name of the profile to be used for this remote endpoint |

Reviewer: ```suggestion
| profile           | reference | The name of the profile to be used for this remote endpoint. |
```

> | remote-id         | string    | The optional remote identifier used during authentication |

Reviewer: ```suggestion
| remote-id         | string    | The optional remote identifier used during authentication. |
```

> | subnet            | ip-prefix | The remote subnet behind the tunnel |

Reviewer: ```suggestion
| subnet            | ip-prefix | The remote subnet behind the tunnel. |
```

> | tunnel-monitor    | container | Properties for monitoring the phase-2 connection. More details [here](plugin_ipsec_client.md#tunnel-monitoring) |

Reviewer: ```suggestion
| tunnel-monitor    | container | Properties for monitoring the phase-2 connection. See [Tunnel Monitoring](plugin_ipsec_client.md#tunnel-monitoring) for more information. |
```

Only one `ipsec-client` can be configured per node, but two `remote`s can be configured per client.
:::

:::warning

> The `ipsec-client > name` cannot start with `ipsec` or `mast`. See notes [here](plugin_ipsec_client.md#release-340)

Reviewer: ```suggestion
The `ipsec-client > name` cannot start with `ipsec` or `mast`. See notes [here](plugin_ipsec_client.md#release-340).
```

| `interval` | Duration (in seconds) of how often to perform an ICMP probe test to the probe-address. |
| `tunnel-monitor-nat-network` | The subnet where traffic originates. The corresponding ingress KNI's fourth octet is used. By default, the subnet `10.128.128.0/28` is used. |

### Configuring services for tunnel traffic

> The user define up to four `ipsec-client > remote` endpoints per node. In addition, the user has to also define the necessary service and service-routes to route the tunnel traffic over the desired WAN interface.

Reviewer: ```suggestion
The user can define up to four `ipsec-client > remote` endpoints per node. In addition, the user must also define the necessary service and service-routes to route the tunnel traffic over the desired WAN interface.
```

```
config

    authority

        router   myRouter
            name           myRouter

            service-route  guest-internet-rte-primary
                name          guest-internet-rte-primary
                service-name  guest-internet

                next-hop      combo-east-1 primary
                    node-name  combo-east-1
                    interface  primary
                exit
            exit

            service-route  guest-internet-rte-secondary
                name          guest-internet-rte-secondary
                service-name  guest-internet

                next-hop      combo-east-1 secondary
                    node-name  combo-east-1
                    interface  secondary
                exit
            exit

        exit

        tenant   guest
            name  guest
        exit

        service  guest-internet
            name           guest-internet
            description    "guest-internet access"
            scope          private
            address        0.0.0.0/0

            access-policy  guest
                source  guest
            exit
        exit
    exit
exit
```

> The `service-route > next-hop > interface` must point to the corresponding `ipsec-client > remote > name`-intf to route the traffic through the tunnel. All the available SSR routing techniques using vectors, service-policy etc can be leverage to define how the untrusted traffic is sent over the tunnels and how to route the traffic. The SSR will route the packets through the appropriate IPSec tunnels and manage the failover as per the policy. On the return path, the encrypted traffic will be processed, decrypted and send back to the client that originates the session.

Reviewer: ```suggestion
The `service-route > next-hop > interface` must point to the corresponding `ipsec-client > remote > name`-intf to route the traffic through the tunnel. All available SSR routing techniques including using vectors, service-policy, etc., can be leveraged to define how the untrusted traffic is sent over the tunnels, and how to route the traffic. The SSR routes the packets through the appropriate IPSec tunnels and manages the failover as per the policy. On the return path, the encrypted traffic is processed, decrypted, and send back to the client that originated the session.
```

## Thirdparty Software & Licenses
- Libreswan v3.23-5.el7_5 (GNU GPLv2)

> ## Troubleshooting

Author: no change to the troubleshooting section. I just moved all the plugin generated config details to the Appendix section which caused this to get bumped up

## Release Notes

### Release 3.4.0

**Release Date:** Jul 13, 2022

> **Router Version** 128T-ipsec-2.2.4-2

Reviewer 2: Router version 2.2.4-2, but release 3.4.0?

Author: plugins have separate versions for conductor and router components. So on field demand we document both version in this format now.

#### New Features and Improvements
- **PLUGIN-1720** Added support for new ciphers for Phase-1 and Phase-2 security

The plugin now support aes-256-gcm, dh-20, sha512 and no-auth encryption schemes.

#### Issues Fixed

- **PLUGIN-880** IPSec client connection is not initiated for certain config combinations

  _**Resolution:**_ The ipsec client names beginning with ipsec and mast are not supported by libreswan. As part of the fix, a new config rule was added to enforce this requirement.

- **PLUGIN-1807** IPSec health monitoring configuration does not report correct status

  _**Resolution:**_ The monitoring scripts were writing data to a non-persistent path resulting in the failure across reboots. The state directory will not be automatically re-generated on system reboots.

- **PLUGIN-1810** Trace logging in configuration results in IPSec failures

>   _**Resolution:**_ Trace level logging is not converted appropriately for IPSec configuration to prevent underlying failures.

Reviewer: ```suggestion
  _**Resolution:**_ Trace level logging was not converted appropriately for IPSec configuration to prevent underlying failures.
```
or "is now converted" The resolution sounds like a restatement of the problem.

Author: I think I mistyped `now` as `not`. It should've said `logging is now converted`.
